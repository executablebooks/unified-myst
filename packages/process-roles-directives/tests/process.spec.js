import { u } from 'unist-builder'
import { processRolesDirectives } from '../src/index.js'

/**
 * @param {any} node
 * @param {any} context
 */
function processRoleSimple(node, context) {
    /** @type any */
    const props = { name: node.name }
    if (context.definitions.size) {
        props.definitions = context.definitions
    }
    if (context.footnotes.size) {
        props.footnotes = context.footnotes
    }
    return [u('element', { name: node.name, ...props })]
}
/**
 * @param {any} node
 */
function processRoleNested(node) {
    if (node.name === 'nested') {
        return [u('leaf')]
    }
    return [u('mystRole', { name: 'nested' })]
}

/**
 * @param {any} node
 */
function processDirectiveSimple(node) {
    return u('mystDirective', node)
}

/**
 * @param {any} node
 */
function processDirectiveNested(node) {
    if (node.name === 'nested') {
        return u('mystDirective', node)
    }
    return u('mystDirective', [
        u('code', {
            lang: '{nested}',
            meta: 'argument',
            value: 'content',
        }),
    ])
}

function processDirectiveRole() {
    return u('mystDirective', [u('mystRole', { name: 'name' })])
}

/**
 * @param {any} node
 * @param {any} context
 */
function processDirectiveDefinition(node, context) {
    if (node.name === 'nested') {
        return u('mystDirective', context)
    }
    return u('mystDirective', [
        u('leaf', context),
        u('code', {
            lang: '{nested}',
            meta: 'argument',
            value: 'content',
        }),
        u('definition', { identifier: 'other' }),
    ])
}

/**
 * @param {any} node
 * @param {any} context
 */
function processDirectiveFootnote(node, context) {
    if (node.name === 'nested') {
        return u('mystDirective', context)
    }
    return u('mystDirective', [
        u('leaf', context),
        u('code', {
            lang: '{nested}',
            meta: 'argument',
            value: 'content',
        }),
        u('footnoteDefinition', { identifier: 'other' }),
    ])
}

describe('Process roles', () => {
    test('null', () => {
        const tree = u('root', [u('paragraph', [u('text', 'Hello')])])
        processRolesDirectives(tree, processRoleSimple, processDirectiveSimple)
        expect(tree).toMatchSnapshot()
    })
    test('basic', () => {
        const tree = u('root', [
            u('paragraph', [u('mystRole', { name: 'name' })]),
        ])
        processRolesDirectives(tree, processRoleSimple, processDirectiveSimple)
        expect(tree).toMatchSnapshot()
    })
    test('multiple', () => {
        const tree = u('root', [
            u('paragraph', [
                u('mystRole', { name: 'name1' }),
                u('mystRole', { name: 'name2' }),
            ]),
            u('paragraph', [u('mystRole', { name: 'name3' })]),
        ])
        processRolesDirectives(tree, processRoleSimple, processDirectiveSimple)
        expect(tree).toMatchSnapshot()
    })
    test('pre-processed', () => {
        const tree = u('root', [
            u('paragraph', [
                u('mystRole', { name: 'name1' }, [u('text', 'Hello')]),
            ]),
        ])
        processRolesDirectives(tree, processRoleSimple, processDirectiveSimple)
        expect(tree).toMatchSnapshot()
    })
    test('nested', () => {
        const tree = u('root', [
            u('paragraph', [u('mystRole', { name: 'name' })]),
        ])
        processRolesDirectives(tree, processRoleNested, processDirectiveSimple)
        expect(tree).toMatchSnapshot()
    })
    test('definition', () => {
        const tree = u('root', [
            u('paragraph', [u('mystRole', { name: 'name' })]),
            u('definition', { identifier: 'test' }),
        ])
        processRolesDirectives(tree, processRoleSimple, processDirectiveSimple)
        expect(tree).toMatchSnapshot()
    })
    test('footnote', () => {
        const tree = u('root', [
            u('paragraph', [u('mystRole', { name: 'name' })]),
            u('footnoteDefinition', { identifier: 'test' }),
        ])
        processRolesDirectives(tree, processRoleSimple, processDirectiveSimple)
        expect(tree).toMatchSnapshot()
    })
})

describe('Process directives', () => {
    test('null', () => {
        const tree = u('root', [
            u('code', { lang: 'name', meta: 'argument', value: 'content' }),
        ])
        processRolesDirectives(tree, processRoleSimple, processDirectiveSimple)
        expect(tree).toMatchSnapshot()
    })
    test('basic', () => {
        const tree = u('root', [
            u('code', { lang: '{name}', meta: 'argument', value: 'content' }),
        ])
        processRolesDirectives(tree, processRoleSimple, processDirectiveSimple)
        expect(tree).toMatchSnapshot()
    })
    test('nested', () => {
        const tree = u('root', [
            u('code', { lang: '{name}', meta: 'argument', value: 'content' }),
        ])
        processRolesDirectives(tree, processRoleSimple, processDirectiveNested)
        expect(tree).toMatchSnapshot()
    })
    test('nested definition', () => {
        const tree = u('root', [
            u('code', { lang: '{name}', meta: 'argument', value: 'content' }),
            u('definition', { identifier: 'test' }),
        ])
        processRolesDirectives(
            tree,
            processRoleSimple,
            processDirectiveDefinition
        )
        expect(tree).toMatchSnapshot()
    })
    test('nested footnote', () => {
        const tree = u('root', [
            u('code', { lang: '{name}', meta: 'argument', value: 'content' }),
            u('footnoteDefinition', { identifier: 'test' }),
        ])
        processRolesDirectives(
            tree,
            processRoleSimple,
            processDirectiveFootnote
        )
        expect(tree).toMatchSnapshot()
    })
    test('role in directive', () => {
        const tree = u('root', [
            u('code', { lang: '{name}', meta: 'argument', value: 'content' }),
        ])
        processRolesDirectives(tree, processRoleSimple, processDirectiveRole)
        expect(tree).toMatchSnapshot()
    })
})
