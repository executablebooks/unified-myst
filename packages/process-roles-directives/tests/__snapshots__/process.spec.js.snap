// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Process directives basic 1`] = `
Object {
  "children": Array [
    Object {
      "meta": "argument",
      "name": "name",
      "position": undefined,
      "type": "mystDirective",
      "value": "content",
    },
  ],
  "type": "root",
}
`;

exports[`Process directives nested 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "meta": "argument",
          "name": "nested",
          "position": undefined,
          "type": "mystDirective",
          "value": "content",
        },
      ],
      "type": "mystDirective",
    },
  ],
  "type": "root",
}
`;

exports[`Process directives nested definition 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "definitions": Set {
            "test",
          },
          "footnotes": Set {},
          "type": "leaf",
        },
        Object {
          "definitions": Set {
            "test",
            "other",
          },
          "footnotes": Set {},
          "type": "mystDirective",
        },
        Object {
          "identifier": "other",
          "type": "definition",
        },
      ],
      "type": "mystDirective",
    },
    Object {
      "identifier": "test",
      "type": "definition",
    },
  ],
  "type": "root",
}
`;

exports[`Process directives nested footnote 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "definitions": Set {},
          "footnotes": Set {
            "test",
          },
          "type": "leaf",
        },
        Object {
          "definitions": Set {},
          "footnotes": Set {
            "test",
            "other",
          },
          "type": "mystDirective",
        },
        Object {
          "identifier": "other",
          "type": "footnoteDefinition",
        },
      ],
      "type": "mystDirective",
    },
    Object {
      "identifier": "test",
      "type": "footnoteDefinition",
    },
  ],
  "type": "root",
}
`;

exports[`Process directives null 1`] = `
Object {
  "children": Array [
    Object {
      "lang": "name",
      "meta": "argument",
      "type": "code",
      "value": "content",
    },
  ],
  "type": "root",
}
`;

exports[`Process directives role in directive 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "name": "name",
              "type": "element",
            },
          ],
          "name": "name",
          "type": "mystRole",
        },
      ],
      "type": "mystDirective",
    },
  ],
  "type": "root",
}
`;

exports[`Process roles basic 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "name": "name",
              "type": "element",
            },
          ],
          "name": "name",
          "type": "mystRole",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "root",
}
`;

exports[`Process roles definition 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "definitions": Set {
                "test",
              },
              "name": "name",
              "type": "element",
            },
          ],
          "name": "name",
          "type": "mystRole",
        },
      ],
      "type": "paragraph",
    },
    Object {
      "identifier": "test",
      "type": "definition",
    },
  ],
  "type": "root",
}
`;

exports[`Process roles footnote 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "footnotes": Set {
                "test",
              },
              "name": "name",
              "type": "element",
            },
          ],
          "name": "name",
          "type": "mystRole",
        },
      ],
      "type": "paragraph",
    },
    Object {
      "identifier": "test",
      "type": "footnoteDefinition",
    },
  ],
  "type": "root",
}
`;

exports[`Process roles multiple 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "name": "name1",
              "type": "element",
            },
          ],
          "name": "name1",
          "type": "mystRole",
        },
        Object {
          "children": Array [
            Object {
              "name": "name2",
              "type": "element",
            },
          ],
          "name": "name2",
          "type": "mystRole",
        },
      ],
      "type": "paragraph",
    },
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "name": "name3",
              "type": "element",
            },
          ],
          "name": "name3",
          "type": "mystRole",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "root",
}
`;

exports[`Process roles nested 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "type": "leaf",
                },
              ],
              "name": "nested",
              "type": "mystRole",
            },
          ],
          "name": "name",
          "type": "mystRole",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "root",
}
`;

exports[`Process roles null 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "type": "text",
          "value": "Hello",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "root",
}
`;

exports[`Process roles pre-processed 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "type": "text",
              "value": "Hello",
            },
          ],
          "name": "name1",
          "type": "mystRole",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "root",
}
`;
